core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -80,114 +146,344 @@
+    * @return the set of characteristics that this store supports
     */
    default Set<Characteristic> characteristics() {
       return EnumSet.noneOf(Characteristic.class);
    }
 
    /**
-    *
-    * @return
+    * Returns a stage that when complete will tell if the current store can be accessed for requests. This can be useful
+    * for some store implementations that may rely on an external source that may become unreachable. This can reduce
+    * sending requests to a store that is not available as future Cache requests will all be met with
+    * {@link StoreUnavailableException}.
+    * <p>
+    * Store availability is polled automatically every so often to update the status if the underlying store changes
+    * availability. This method will not be invoked concurrently with itself (ie. this method will not be invoked
+    * until after the previous stage has completed), but will be invoked concurrently with other operations, excluding
+    * start and stop.
+    * <p>
+    * The other benefit of implementing this method is if the store is configured to be
+    * {@link StoreConfiguration#async()}. This allows for this store to be unavailable for short periods of time
+    * until its buffer fills.
+    * @implSpec
+    * The default implementations returns a completed stage with the value {@code Boolean.TRUE}
+    * @return stage that when complete signals if the store is available
     */
    default CompletionStage<Boolean> isAvailable() {
       return CompletableFutures.completedTrue();
    }
 
    /**
-    *
-    * @param segment
-    * @param key
-    * @return
+    * Returns a stage that will contain the value loaded from the store. If a {@link MarshallableEntry} needs to be
+    * created here, {@link InitializationContext#getMarshallableEntryFactory()} ()} and {@link
+    * InitializationContext#getByteBufferFactory()} should be used to assist.
+    * <p>
+    * <h4>Summary of Characteristics Effects</h4>
+    * <table border="1" cellpadding="1" cellspacing="1" summary="Summary of Characteristics Effects">
+    *    <tr>
+    *       <th bgcolor="#CCCCFF" align="left">Characteristic</th>
+    *       <th bgcolor="#CCCCFF" align="left">Effect</th>
+    *    </tr>
+    *    <tr>
+    *       <td valign="top">{@link Characteristic#WRITE_ONLY}</td>
+    *       <td valign="top">This method will never be invoked</td>
+    *    </tr>
+    *    <tr>
+    *       <td valign="top">{@link Characteristic#EXPIRATION}</td>
+    *       <td valign="top">When set this method must not return expired entries</td>
+    *    </tr>
+    *    <tr>
+    *       <td valign="top">{@link Characteristic#SEGMENTABLE}</td>
+    *       <td valign="top">When this is not set the provided {@code segment} parameter may be ignored</td>
+    *    </tr>
+    * </table>
+    * <p>
+    * If any problem is encountered, it is recommended that the store wrap any such Throwable in a PersistenceException
+    * completing the stage exceptionally.
```

Comment:
"If a problem is encountered, it is recommended to wrap any created/caught throwables in a {@link PersistenceException} and the stage be completed exceptionally.

Reply:
This appers for most methods, so I won't duplicate the comment.